### ISSUE TITLE
Planning to write detect / other functions?

### ISSUE THREAD

**Issue by vsoch** [archspec/archspec-go#4]
Hey Archspec,

I noticed that there aren't functions here for detect, etc., is this something you are planning to write, to be akin to the Python module? Thanks!

**Comment by alalazo** [comment]
Yes, this project currently is unfinished but the idea is to have feature parity with the Python API.

**Comment by vsoch** [comment]
Ah fantastic! I guess I'll subscribe for updates/notifications then. I'm working on a sibling library called containerspec that I'd like to use it with.

**Comment by vsoch** [comment]
hey @alalazo I'd really like to use this, any updates / plans for working on it?

**Comment by alalazo** [comment]
@vsoch Would you have time to help with implementing the command line tools? This has been since ever on my TODO list at very low priority... :sweat:

**Comment by vsoch** [comment]
I think I tried last August (2021) but I didn't know how to go about it. E.g., I know you can look at runtime.GOOS == "linux" but I wasn't sure about your plan to actually match to a specific arch/architecture. Do you have a design document or similar somewhere?

**Comment by vsoch** [comment]
Ok so even if you don’t know details of the implementation, do you have a general design idea? I’m just not familiar with how archspec works (or is supposed to) beyond doing the detecting. Is that something that needs to be figured out too? How does it work in the Python version?

**Comment by vsoch** [comment]
Have you seen https://chromium.googlesource.com/external/github.com/klauspost/cpuid/+/30e450625213468c7ec31fec6102dab6504adcc9/cpuid.go ?

**Comment by vsoch** [comment]
Here is the repository (linked at the top there): https://github.com/klauspost/cpuid what are the main differences between the library there and archspec? Is it just that archspec has a comparison / dag thing going on? The question I'm interested in is if we can use that library - e.g., is it possible to match results between there and the json here, and if not, what would we do if we parse some unknown microarchitecture?

**Comment by alalazo** [comment]
> what are the main differences between the library there and archspec? 

That library seems focused on hardware detection, so I'd expect it to give the same level of details as e.g. `/proc/cpuinfo` and to be unable to "reason" about architecture compatibility. `archspec` main focus is on binary compatibility, so it provides a few human-readable labels (e.g. `bulldozer` or `graviton`) and can "reason" about any of the architecture in the JSON file, while it can't distinguish two different chips that implement the same microarchitecture.

I think we can use the library you linked as an alternate method of detecting "raw" information on the current host and matching it to a known target in the JSON file.

**Comment by vsoch** [comment]
Gotcha! Yeah I can mess around with that!   I will make some time soon.

**Comment by ChristianKniep** [comment]
Any progress on that issue? I am wondering myself how to make use of this package? The [archspec-feature-discovery](https://github.com/archspec/archspec-feature-discovery/blob/master/main.go#L74) tool seems to execute the python version.

**Comment by vsoch** [comment]
I was interested for this tool I was working on https://github.com/vsoch/containerspec but I think I lost interest to work on it (August 2021!).
